src/Analyzer/InDepthQueryTreeVisitor.h

```diff
@@ -84,4 +91,144 @@ class InDepthQueryTreeVisitor
 template <typename Derived>
 using ConstInDepthQueryTreeVisitor = InDepthQueryTreeVisitor<Derived, true /*const_visitor*/>;
 
+/** Same as InDepthQueryTreeVisitor and additionally keeps track of current scope context.
+  * This can be useful if your visitor has special logic that depends on current scope context.
+  */
+template <typename Derived, bool const_visitor = false>
+class InDepthQueryTreeVisitorWithContext
+{
+public:
+    using VisitQueryTreeNodeType = std::conditional_t<const_visitor, const QueryTreeNodePtr, QueryTreeNodePtr>;
+
+    explicit InDepthQueryTreeVisitorWithContext(ContextPtr context)
+        : current_context(std::move(context))
+    {}
+
+    /// Return true if visitor should traverse tree top to bottom, false otherwise
+    bool shouldTraverseTopToBottom() const
+    {
+        return true;
+    }
+
+    /// Return true if visitor should visit child, false otherwise
+    bool needChildVisit(VisitQueryTreeNodeType & parent [[maybe_unused]], VisitQueryTreeNodeType & child [[maybe_unused]])
+    {
+        return true;
+    }
+
+    const ContextPtr & getContext() const
+    {
+        return current_context;
+    }
+
+    const ContextPtr & getContext()
```

Comment:
You don't need it, because `const ContextPtr & getContext() const` can be called from a mutable object.

Reply:
Aggreed. Updated.